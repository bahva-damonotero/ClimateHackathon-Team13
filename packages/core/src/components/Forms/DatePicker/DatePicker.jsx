import React, { useEffect, useRef, useState } from 'react';
import PropTypes from 'prop-types';
import datePicker from '@uswds/uswds/js/usa-date-picker';

import { FormError, Label } from '..';
import { Hint } from '../..';

export default function DatePicker({
  children,
  disabled,
  error,
  hint,
  id,
  maxDate,
  minDate,
  onChange,
  rangeDate,
  value,
  ...props
}) {
  const labelId = `${id}-label`;
  const hintId = `${id}-hint`;

  const containerRef = useRef();
  const datePickerRef = useRef();
  const inputRef = useRef();

  const [ready, setReady] = useState(false);

  // initialization
  useEffect(() => {
    const containerEl = containerRef.current;
    const inputEl = inputRef.current;
    if (!containerEl || !inputEl) return;

    // call USWDS's `on` function to initialize the component
    datePicker.on(containerEl);
    setReady(true);

    // hook into the input's `change` event to monitor changes
    inputEl.addEventListener('change', (e) => {
      if (onChange) onChange(e);
    });

    // on dismount, call USWDS's `off` function to cleanup the component
    return () => {
      datePicker.off(containerEl);
    };
  }, []);

  // call USWDS's `setCalendarValue` when the value changes
  // outside of this component
  useEffect(() => {
    const hiddenInputEl = inputRef.current;
    const datePickerEl = datePickerRef.current;
    if (!hiddenInputEl || !datePickerEl) return;

    if (value !== hiddenInputEl.value) {
      if (!value) {
        // reset the date picker
        const shownInputEl = datePickerEl.querySelector(
          '.usa-date-picker__external-input'
        );
        shownInputEl.value = '';
        hiddenInputEl.value = '';
      } else {
        datePicker.setCalendarValue(datePickerEl, value);
      }
    }
  }, [value, inputRef, datePickerRef]);

  // call USWDS's `enable` or `disable` function when the disabled prop changes
  useEffect(() => {
    const datePickerEl = datePickerRef.current;
    if (!datePickerEl) return;

    const disabledState = datePickerEl.querySelector(`#${id}`).disabled;
    if (disabled === disabledState) return; // the input already has the right disabled state

    if (disabled) datePicker.disable(datePickerEl);
    else datePicker.enable(datePickerEl);
  }, [disabled]);

  return (
    <div
      className="bah-uswds-date-picker"
      style={{ display: ready ? 'block' : 'none' }}
      ref={containerRef}
    >
      <Label error={error} id={labelId} target={id}>
        {children}
      </Label>

      <FormError error={error} />

      <Hint id={hintId}>{hint}</Hint>

      <div
        className="usa-date-picker"
        data-default-value={value}
        data-max-date={maxDate}
        data-min-date={minDate}
        data-range-date={rangeDate}
        ref={datePickerRef}
      >
        <input
          aria-describedby={hintId}
          aria-labelledby={labelId}
          className="usa-input"
          disabled={disabled}
          id={id}
          name={id}
          ref={inputRef}
          type="text"
          {...props}
        />
      </div>
    </div>
  );
}

DatePicker.propTypes = {
  /** Label for input */
  children: PropTypes.node.isRequired,
  /** Should the date picker be disabled? */
  disabled: PropTypes.bool,
  /** Error message to display */
  error: PropTypes.string,
  /** Hint to be shown for the date picker */
  hint: PropTypes.node,
  /** ID of the input */
  id: PropTypes.string.isRequired,
  /** The date picker will not allow a date selection after this date. Must be in the format: `YYYY-MM-DD` */
  maxDate: PropTypes.string,
  /** The date picker will not allow a date selection before this date. Must be in the format: `YYYY-MM-DD` */
  minDate: PropTypes.string,
  /** Function called when the date is changed */
  onChange: PropTypes.func,
  /** When set, the date picker will show a range selection from the range date. Must be in the format: `YYYY-MM-DD` */
  rangeDate: PropTypes.string,
  /** Used to set the value of the input. Must be in the format: `YYYY-MM-DD` */
  value: PropTypes.string
};

DatePicker.defaultProps = {
  disabled: false,
  hint: 'mm/dd/yyyy'
};
