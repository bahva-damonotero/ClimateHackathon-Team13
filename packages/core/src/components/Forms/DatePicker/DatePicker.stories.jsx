import React from 'react';
import { action } from '@storybook/addon-actions';

import Form from '../Form';
import DatePicker from '.';

export default {
  title: 'Forms/DatePicker',
  component: DatePicker
};

const Template = (args) => (
  <Form>
    <DatePicker {...args} />
  </Form>
);

const defaultArgs = {
  children: 'Label',
  id: 'datepicker-example',
  onChange: action('onChange')
};

export const def = Template.bind({});
def.storyName = 'Default';
def.args = {
  ...defaultArgs
};

export const initialValue = Template.bind({});
initialValue.args = {
  ...defaultArgs,
  value: '2020-01-01'
};

export const minAndMaxDate = Template.bind({});
minAndMaxDate.args = {
  ...defaultArgs,
  minDate: '1995-10-25',
  maxDate: '1999-12-12'
};

export const dateRange = Template.bind({});
dateRange.storyName = 'Date range (different from Date Range Picker)';
dateRange.args = {
  ...defaultArgs,
  rangeDate: '2020-10-08'
};
