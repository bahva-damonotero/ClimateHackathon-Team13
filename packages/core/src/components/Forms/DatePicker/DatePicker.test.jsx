import React from 'react';
import { createEvent, fireEvent, render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import datePicker from '@uswds/uswds/js/usa-date-picker';

import DatePicker from '.';

describe('DatePicker', () => {
  const id = 'date-picker-example';
  const label = 'Label';

  const INPUT_SELECTOR = '.usa-date-picker__internal-input';
  const EXTERNAL_INPUT_SELECTOR = '.usa-date-picker__external-input';

  it('should call `datePicker.on()` to initialize the component', () => {
    const onSpy = jest.spyOn(datePicker, 'on');

    render(<DatePicker id={id}>{label}</DatePicker>);

    expect(onSpy).toHaveBeenCalled();
  });

  it('should call `onChange` when the value is changed', () => {
    const onChange = jest.fn();

    const { container } = render(
      <DatePicker id={id} onChange={onChange}>
        {label}
      </DatePicker>
    );

    const inputNode = container.querySelector(INPUT_SELECTOR);

    const event = createEvent('change', inputNode, {
      target: { value: '1995-10-25' }
    });
    fireEvent(inputNode, event);

    expect(onChange).toHaveBeenCalledWith(event);
  });

  it('should be able to update the value programmatically', () => {
    // render the date picker with no value
    const initialProps = {
      id,
      onChange: () => {}
    };

    const { container, rerender } = render(
      <DatePicker {...initialProps}>{label}</DatePicker>
    );

    const inputNode = container.querySelector(INPUT_SELECTOR);
    expect(inputNode.value).toBe('');

    // update the date picker's value programmatically/externally
    const newValue = '1995-10-25';

    rerender(
      <DatePicker {...initialProps} value={newValue}>
        {label}
      </DatePicker>
    );

    expect(inputNode.value).toBe(newValue);
  });

  it('should be be able to clear the value programmatically', () => {
    // render the date picker with a value
    const initialValue = '1995-10-25';
    const initialProps = {
      id,
      onChange: () => {},
      value: initialValue
    };

    const { container, rerender } = render(
      <DatePicker {...initialProps}>{label}</DatePicker>
    );

    const hiddenInputNode = container.querySelector(INPUT_SELECTOR);
    const shownInputNode = container.querySelector(EXTERNAL_INPUT_SELECTOR);

    expect(hiddenInputNode.value).toBe(initialValue);
    expect(shownInputNode.value).toBe('10/25/1995');

    // clear the date picker's value programmatically/externally
    const newValue = undefined;

    rerender(
      <DatePicker {...initialProps} value={newValue}>
        {label}
      </DatePicker>
    );

    expect(hiddenInputNode.value).toBe('');
    expect(shownInputNode.value).toBe('');
  });

  it('changing the `disabled` prop to true should call `datePicker.disable()`', () => {
    const disableSpy = jest.spyOn(datePicker, 'disable');

    const { rerender } = render(<DatePicker id={id}>{label}</DatePicker>);

    rerender(
      <DatePicker id={id} disabled>
        {label}
      </DatePicker>
    );

    expect(disableSpy).toHaveBeenCalled();
  });

  it('changing the `disabled` prop to false should call `datePicker.enable()`', () => {
    const enableSpy = jest.spyOn(datePicker, 'enable');

    const { rerender } = render(
      <DatePicker id={id} disabled>
        {label}
      </DatePicker>
    );

    rerender(<DatePicker id={id}>{label}</DatePicker>);

    expect(enableSpy).toHaveBeenCalled();
  });

  it('should match snapshot', () => {
    const tree = renderer
      .create(<DatePicker id={id}>{label}</DatePicker>)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
