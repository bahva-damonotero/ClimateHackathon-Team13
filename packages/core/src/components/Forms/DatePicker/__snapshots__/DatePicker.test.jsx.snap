// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePicker should match snapshot 1`] = `
<div
  className="bah-uswds-date-picker"
  style={
    Object {
      "display": "none",
    }
  }
>
  <label
    className="usa-label"
    htmlFor="date-picker-example"
    id="date-picker-example-label"
  >
    Label
  </label>
  <span
    className="usa-hint"
    id="date-picker-example-hint"
  >
    mm/dd/yyyy
  </span>
  <div
    className="usa-date-picker"
  >
    <input
      aria-describedby="date-picker-example-hint"
      aria-labelledby="date-picker-example-label"
      className="usa-input"
      disabled={false}
      id="date-picker-example"
      name="date-picker-example"
      type="text"
    />
  </div>
</div>
`;
